File: shared/errors/codes.go
// Package apperrors defines unified error codes shared between the
// Go server and the Dart client.
package apperrors

import "net/http"

// ErrorCode is a numeric error code sent to clients in error messages.
type ErrorCode int

const (
	// ErrOK is the zero value, meaning no error.
	ErrOK ErrorCode = iota
	// ErrInvalidToken means the authentication token is missing, expired, or invalid.
	ErrInvalidToken
	// ErrUserNotFound means the user does not exist.
	ErrUserNotFound
	// ErrAlreadyInCall means the user is already in an active call.
	ErrAlreadyInCall
	// ErrNotInPool means the user is not in any matching pool.
	ErrNotInPool
	// ErrPoolTimeout means the user waited too long in the pool.
	ErrPoolTimeout
	// ErrCallNotFound means the call ID does not exist.
	ErrCallNotFound
	// ErrCallRejected means the callee rejected the call.
	ErrCallRejected
	// ErrNetwork means a network-level failure occurred.
	ErrNetwork
	// ErrInternal means an internal server error.
	ErrInternal
	// ErrRateLimited means the client sent too many requests.
	ErrRateLimited
	// ErrBlacklisted means the caller and callee have blocked each other.
	ErrBlacklisted
	// ErrInvalidParams means the request parameters are invalid.
	ErrInvalidParams
)

// codeMessages maps each ErrorCode to a human-readable message.
var codeMessages = map[ErrorCode]string{
	ErrOK:            "ok",
	ErrInvalidToken:  "invalid or expired token",
	ErrUserNotFound:  "user not found",
	ErrAlreadyInCall: "user is already in a call",
	ErrNotInPool:     "user is not in any matching pool",
	ErrPoolTimeout:   "matching pool wait timeout",
	ErrCallNotFound:  "call not found",
	ErrCallRejected:  "call was rejected",
	ErrNetwork:       "network error",
	ErrInternal:      "internal server error",
	ErrRateLimited:   "too many requests",
	ErrBlacklisted:   "user is blacklisted by peer",
	ErrInvalidParams: "invalid request parameters",
}

// codeHTTPStatus maps each ErrorCode to an HTTP status code.
var codeHTTPStatus = map[ErrorCode]int{
	ErrOK:            http.StatusOK,
	ErrInvalidToken:  http.StatusUnauthorized,
	ErrUserNotFound:  http.StatusNotFound,
	ErrAlreadyInCall: http.StatusConflict,
	ErrNotInPool:     http.StatusBadRequest,
	ErrPoolTimeout:   http.StatusRequestTimeout,
	ErrCallNotFound:  http.StatusNotFound,
	ErrCallRejected:  http.StatusConflict,
	ErrNetwork:       http.StatusBadGateway,
	ErrInternal:      http.StatusInternalServerError,
	ErrRateLimited:   http.StatusTooManyRequests,
	ErrBlacklisted:   http.StatusForbidden,
	ErrInvalidParams: http.StatusBadRequest,
}

// Message returns the human-readable message for the error code.
func (c ErrorCode) Message() string {
	if msg, ok := codeMessages[c]; ok {
		return msg
	}
	return "unknown error"
}

// HTTPStatus returns the recommended HTTP status code for the error.
func (c ErrorCode) HTTPStatus() int {
	if status, ok := codeHTTPStatus[c]; ok {
		return status
	}
	return http.StatusInternalServerError
}

// Int returns the integer representation of the error code.
func (c ErrorCode) Int() int {
	return int(c)
}

File: shared/lib/errors/codes.dart
/// Unified error codes shared between server and client.
///
/// Mirrors `shared/errors/codes.go` on the Go side.
library;

/// Error code constants.
class ErrorCode {
  static const int ok = 0;
  static const int invalidToken = 1;
  static const int userNotFound = 2;
  static const int alreadyInCall = 3;
  static const int notInPool = 4;
  static const int poolTimeout = 5;
  static const int callNotFound = 6;
  static const int callRejected = 7;
  static const int network = 8;
  static const int internal = 9;
  static const int rateLimited = 10;
  static const int blacklisted = 11;
  static const int invalidParams = 12;
}

/// Human-readable messages for each error code.
const Map<int, String> errorMessages = {
  ErrorCode.ok: 'ok',
  ErrorCode.invalidToken: 'invalid or expired token',
  ErrorCode.userNotFound: 'user not found',
  ErrorCode.alreadyInCall: 'user is already in a call',
  ErrorCode.notInPool: 'user is not in any matching pool',
  ErrorCode.poolTimeout: 'matching pool wait timeout',
  ErrorCode.callNotFound: 'call not found',
  ErrorCode.callRejected: 'call was rejected',
  ErrorCode.network: 'network error',
  ErrorCode.internal: 'internal server error',
  ErrorCode.rateLimited: 'too many requests',
  ErrorCode.blacklisted: 'user is blacklisted by peer',
  ErrorCode.invalidParams: 'invalid request parameters',
};

/// Returns the human-readable message for [code].
String errorMessage(int code) =>
    errorMessages[code] ?? 'unknown error';
